src/include/migraphx/iterator.hpp

```diff
@@ -203,7 +215,16 @@ struct iterator_operators
     template <class I, class U = T>
     constexpr auto operator[](I n) const -> decltype(*(static_cast<const U&>(*this) + n))
     {
-        return *(static_cast<const U&>(*this) + n);
+        auto it = static_cast<const U&>(*this) + n;
+        // cppcheck-suppress migraphx-ConditionalAssert
+        if constexpr(std::is_reference<decltype(*it)>{})
+        {
+            // Ensure that result is not an internal reference
+            assert((bit_cast<std::uintptr_t>(&*it) < bit_cast<std::uintptr_t>(&it) or
```

Comment:
Just for sanity sake add a unit test for these assert cases

Reply:
There is already unit tests for this.

For random access iterators, we dont allow internal references(an iterator should just return by value in this case). Boost.Iterator returns a `operator_brackets_proxy` but it just caused a lot more problems that the newer libraries like ranges-v3 or Boost.StlInterface just nuked it(see [here](https://github.com/ericniebler/range-v3/commit/e510379b2208c1e1697ae52287c53fd29ada753f) and [here](https://github.com/ericniebler/range-v3/issues/197)).

However, if you do return an internal reference you can get a dangling reference that might be hard to diagnose so I add an assert with a message to make it a little clearer that internal references aren't supported for this case.